### PR TITLE
Use AzureDevOps counter expression to generate CI version instead of using a text file on the file-share

### PR DESCRIPTION
**Pull request by kartheekp-ms**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/258

Regression? Last working version:

## Description

The Semantic Version for NuGet assemblies/Packages is generated [here](https://github.com/NuGet/NuGet.Client/blob/dev/build/config.props#L18). There is a dependency on fileshare (\\ddfiles) to generate CI version. I am sure there could be other ways to fix the linked issue. This PR suggests an approach to increment the CI version using [counter expression](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#counter) in Azure DevOps. I added a new job the CI pipeline which could increase CI duration by a minute due to the dependency on [fetching semantic version from `config.props`](https://github.com/NuGet/NuGet.Client/blob/dev/build/config.props#L18) using `msbuild.exe`.

I have defined `prefix` of `counter expression` on `build definition name (official or private builds)` and `semantic version` so that the CI version increments separately for official or private builds. As of this minute, the official build CI version is 7121. The `seed` starts with `7130` and increments by 1 separately for official or private builds. The following tables displays CI version number for next three builds after this PR have been merged to `dev` and all the team members rebase their branches with `dev`

Build Definition   Name | Build Number | CI Version
-- | -- | --
NuGet-Client Private Build | 1 | 7132 (I ran 2 private builds hence CI version starts with 7132)
NuGet-Client Private Build | 2 | 7133
NuGet-Client Private Build | 3 | 7134

Build Definition   Name | Build Number | CI Version
-- | -- | --
NuGet-Client Official Build | 1 | 7130
NuGet-Client Official Build | 2 | 7131
NuGet-Client Official Build | 3 | 7132

This approach ensures that for every CI build, version number of binaries is higher than previous version of the product. After we merge this PR into `dev` and cherry-pick these commits onto our servicing branches then I will create a follow up PR to change the seed back to 0 so that CI revision number starts from `0` for any change in major, minor or patch versions.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] N/A 

- **Documentation**
  
  - [x] N/A

### CODE REVIEW

File: eng/pipelines/templates/pipeline.yml
Comment (donnie-msft) [inline eng/pipelines/templates/pipeline.yml:18]: 7130 is a seed value? Do we need to move this higher once your PR is ready to merge? Can it be a build variable instead?

**Review by donnie-msft — COMMENTED** [review]
I like the new step! 
Private and Dev will be separate builds, but the build numbers will overlap now, right?
You have 7132 listed twice in the table, but they represent completely different builds (private and official). I'm not sure that's good, as it could be easy to confuse builds? Should we have different seeds, maybe with Private being much a higher value as we did previously?

File: eng/pipelines/templates/pipeline.yml
Comment (donnie-msft) [reply eng/pipelines/templates/pipeline.yml:18]: I see, you will have a followup PR...

**Comment by kartheekp-ms** [comment]
> Private and Dev will be separate builds, but the build numbers will overlap now, right?
You have 7132 listed twice in the table, but they represent completely different builds (private and official). I'm not sure that's good, as it could be easy to confuse builds? Should we have different seeds, with Private being much a higher value as we did previously?

Yes, the build numbers will overlap. I guess it should be okay because the `private and official` build artifacts will be dropped to separate locations on the file-share. The lowest private build counter for `5.10.0` version is [`14370`](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4437101&view=results). If we merge this PR then the CI version for private builds will start from `7132`. AFAIK, the chances of private build number overlapping for the same semantic version (i.e. `5.10.0`) are slim because it requires `7238` private builds for the build number to be again `5.10.0.14370`. Once we implement [this epic](https://github.com/NuGet/Client.Engineering/issues/313), we no longer publish build artifacts to the file share from the build pipeline.

Andy suggested Arcade SDK versioning strategy as another option [here](https://github.com/nuget/client.engineering/issues/258#issuecomment-611203310). I liked the versioning story, but it may take some time for me to understand it thoroughly, write a spec, review with the team, and implementation. Since we have stop relying on the file share for builds by end of this month, I thought the changes suggested in this PR would work and unblock us.

I am happy to tweak the logic based on the feedback.

File: eng/pipelines/templates/pipeline.yml
Comment (zivkan) [inline eng/pipelines/templates/pipeline.yml:18]: > I added a new job the CI pipeline which could increase CI duration by a minute due to the dependency on fetching semantic version from config.props using msbuild.exe.

This could be a LOT more than 1 minute if all the agents in the pool are busy.

Rather than making these job variables, can you define the variable in the step definition, and therefore move the "Set Semantic Version" to the same job as where it's used?

Comment (zivkan) [inline eng/pipelines/templates/pipeline.yml:18]: >  This approach ensures that for every CI build, version number of binaries is higher than previous version of the product. After we merge this PR into dev and cherry-pick these commits onto our servicing branches then I will create a follow up PR to change the seed back to 0 so that CI revision number starts from 0 for any change in major, minor or patch versions.

What's the plan for old releases if we have to service them?

Do we cherry-pick this PR's commit, but not the new PR's commit which changes the default to zero?

Comment (kartheekp-ms) [reply eng/pipelines/templates/pipeline.yml:18]: I forgot to mention in the PR description, but I tried several ways to keep `Set Semantic Version` in `Initialize_Build` job instead of creating a new job. I couldn't figure out a best way because [doc](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#counter) says `counter function can only be used in an expression that defines a variable`.  This [doc](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#user-defined-variables) says, In YAML pipelines, user-defined variables can set variables at the `root, stage, and job` level and in the UI. We execute a [PowerShell script in CI pipeline](https://github.com/NuGet/NuGet.Client/blob/dev/eng/pipelines/templates/Initialize_Build.yml#L43) to get the semantic version from [config.props](https://github.com/NuGet/NuGet.Client/blob/dev/build/config.props#L18). To satisfy the requirement for using `counter` expression, I ended up creating a new CI job and set a dependency for `Initialize_Build` job.

AFAIK, if we define `semantic version` as a variable in the `.yaml` file instead of reading from `config.props` which eliminates the need to execute a PowerShell script to fetch it, then we can remove `Intialize_Build_SemanticVersion` job completely.

Comment (kartheekp-ms) [reply eng/pipelines/templates/pipeline.yml:18]: One option is to cherry-pick this commit on the need basis. Second option is to cherry-pick this commit proactively into all the service branches after merging into `dev`. I lean towards second option so that we don't have any last-minute surprises. I think, we must cherry-pick other dependent commits such as https://github.com/NuGet/NuGet.Client/pull/3883 and https://github.com/NuGet/NuGet.Client/pull/3890 (may be but not sure)

Comment (zivkan) [reply eng/pipelines/templates/pipeline.yml:18]: ok, makes sense. I thought/hoped it was possible to define variables at the task level, but it does make sense that a "job" is the lowest level it can be set.

I believe with your change that neither your new job, nor the existing "initialize build" job need the fileshare, so can you please try changing `name: VSEng-MicroBuildVS2019` to `vmImage: windows-latest` to see if it works on a "hosted" agent, so we're not affected when the microbuild pool is full.

Comment (kartheekp-ms) [reply eng/pipelines/templates/pipeline.yml:18]: Changed `Initialize_Build_SemanticVersion` and `Initialize_Build` jobs to run on `vmImage: windows-latest`.  These two jobs have passed on the [CI with build number 5.10.0.7132](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4473701&view=results) and other jobs are in-progress.

**Review by zivkan — APPROVED** [review]

**Pull request merged**
